templates/llm-router/README.ipynb

```diff
@@ -0,0 +1,1038 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Background\n",
+    "\n",
+    "Whenever we use an LLM we would like to get the highest response quality but are often restricted to a limited cost budget. Closed models, such as GPT-4, are known to be the highest quality models, but they can get very expensive especially when running them on a very number of queries. On the other hand, OSS models can be much cheaper, but their responses may not be of the same quality, especially for complex or domain-specific queries.\n",
+    "\n",
+    "The goal of this tutorial is to show you how you can train a \"smart router\", i.e. a model that can dynamically decide, based on the query text, whether to call a closed model or an OSS model. Here's a schematic view of a smart router:\n",
+    "![Smart Router](assets/router_schema.png)\n",
+    "\n",
+    "The goal is to learn a smart router such that, given a set of queries, we can get responses with the highest overall quality while minimizing the total cost.\n",
+    "\n",
+    "# Approach\n",
+    "\n",
+    "We are going to train a classifier to decide, based only on the query text, whether to route the query to an OSS model vs. a closed one. In this tutorial, we will make the following design choices: \n",
+    "1. We will quantify a response quality on a scale of `[1, 5]` (5-star).\n",
+    "2. For simplicity, we will assume that the closed always achieves 5-start quality. \n",
+    "3. We will use GPT-4 as a representative of closed models and Mixtral 8x7B for OSS models.\n",
+    "\n",
+    "More concurrently, let us assume that closed models have perfect a quality (5/5 score). our goal is to reduce cost significantly (say by 50%) while maintaining a high overal quality (4.8/5 score).\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Implementation overview\n",
+    "\n",
+    "### 1) Generate labeled data\n",
```

Comment:
### 1) Generate labeled data
The label for such classifier does not exist, and we explore one way to generate such data for a specific use-case.

Reply:
add a glimpse of explanation to each section ^